### PR TITLE
Allow dynamic known types

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we allow appending the system's known types at runtime in certain situations.

Such operation may only be performed in a compile-time utility and should never be done in a user's production code.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: base/src/main/java/io/spine/type/KnownTypes.java [hunk 1]
```diff
@@ -169,4 +172,51 @@ private ClassName get(TypeUrl typeUrl) {
...
+         * Extends the known types with some more types.
+         *
+         * <p>This method should never be called in a client code. The sole purpose of extending
+         * the known types is for running compile-time checks on the user types.
+         *
+         * @throws java.lang.SecurityException if called from the client code
+         */
+        @Internal
+        public static void extendWith(TypeSet moreKnownTypes) {
+            InvocationGuard.allowOnly("io.spine.typehack.MoreKnownTypes");
```
Comment (alexander-yevsyukov) [inline base/src/main/java/io/spine/type/KnownTypes.java]: Can we avoid a hack being exposed in a separate package, please?

File: base/src/main/java/io/spine/type/KnownTypes.java [hunk 2]
```diff
@@ -20,61 +20,65 @@
...
  * the {@linkplain io.spine.code.proto.FileDescriptors#KNOWN_TYPES resource file}.
- *
- * @author Mikhail Mikhaylov
- * @author Alexander Litus
- * @author Alexander Yevsyukov
- * @author Dmytro Dashenkov
  */
 @Internal
-public enum KnownTypes {
+@Immutable
```
Comment (alexander-yevsyukov) [inline base/src/main/java/io/spine/type/KnownTypes.java]: Please document the limitations of immutability.

File: tools/plugin-base/src/main/java/io/spine/typehack/package-info.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * This package contains classes used to hack the Spine type system for the needs of
+ * the compile-time checkers and processors.
+ *
+ * <p>DO NOT use these classes, as well as any other classes of the {@code plugin-base} module in
```
Comment (alexander-yevsyukov) [inline tools/plugin-base/src/main/java/io/spine/typehack/package-info.java]: Can we have `@Internal` on the package?

File: tools/plugin-base/src/main/java/io/spine/typehack/MoreKnownTypes.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import io.spine.code.proto.TypeSet;
+import io.spine.type.KnownTypes;
+
+import java.io.File;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A utility for extending the system {@link KnownTypes}.
```
Comment (alexander-yevsyukov) [inline tools/plugin-base/src/main/java/io/spine/typehack/MoreKnownTypes.java]: Please provide more information on why this is all needed. With such dry notes it looks strange.

File: tools/plugin-base/src/main/java/io/spine/typehack/MoreKnownTypes.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.typehack;
```
Comment (alexander-yevsyukov) [inline tools/plugin-base/src/main/java/io/spine/typehack/MoreKnownTypes.java]: Can we think of another package name? If it's part of the build process, why doesn't it go under a corresponding plug-in package?

File: tools/plugin-base/src/main/java/io/spine/typehack/package-info.java (hunk 3 above)
Comment (dmdashenkov) [reply tools/plugin-base/src/main/java/io/spine/typehack/package-info.java]: We do now.

File: base/src/main/java/io/spine/type/KnownTypes.java (hunk 2 above)
Comment (dmdashenkov) [reply base/src/main/java/io/spine/type/KnownTypes.java]: Done.

**Pull request merged**